pkg/cloudprovider/providers/azure/azure_zones.go

```diff
@@ -61,14 +60,31 @@ func (az *Cloud) GetZone() (cloudprovider.Zone, error) {
 // This is particularly useful in external cloud providers where the kubelet
 // does not initialize node data.
 func (az *Cloud) GetZoneByProviderID(providerID string) (cloudprovider.Zone, error) {
-	return cloudprovider.Zone{}, errors.New("GetZoneByProviderID not implemented")
+	nodeName, err := splitProviderID(providerID)
+	if err != nil {
+		return cloudprovider.Zone{}, err
+	}
+	return az.GetZoneByNodeName(nodeName)
 }
 
 // GetZoneByNodeName implements Zones.GetZoneByNodeName
 // This is particularly useful in external cloud providers where the kubelet
 // does not initialize node data.
 func (az *Cloud) GetZoneByNodeName(nodeName types.NodeName) (cloudprovider.Zone, error) {
-	return cloudprovider.Zone{}, errors.New("GetZoneByNodeName not imeplemented")
+	vm, exists, err := az.getVirtualMachine(nodeName)
+
+	if err != nil {
+		return cloudprovider.Zone{}, err
+	}
+
+	if !exists {
+		return cloudprovider.Zone{}, nil
+	}
+	zone := cloudprovider.Zone{
+		FailureDomain: *(vm.AvailabilitySet.ID),
```

Comment:
This is a full resource ID for the availability set, but fault domain should be a number.
Should be:
VirtualMachineProperties/VirtualMachineInstanceView/PlatformFaultDomain

Reference:
https://docs.microsoft.com/en-us/azure/virtual-machines/windows/manage-availability
https://github.com/kubernetes/kubernetes/blob/74f4fda7beba6d6f9b56f92b08c3b11cd27397d4/pkg/cloudprovider/providers/azure/azure_zones.go#L53

Reply:
thanks for the hint.